src/esp/gfx/Renderer.cpp

```diff
@@ -23,109 +23,29 @@ namespace esp {
 namespace gfx {
 
 struct Renderer::Impl {
-  Impl(int width, int height)
-      : framebufferSize_(width, height),
-        colorBuffer_(),
-        depthBuffer_(),
-        objectIdBuffer_(),
-        depthRenderbuffer_(),
-        framebuffer_({{}, framebufferSize_}) {
+  Impl() {
     GL::Renderer::enable(GL::Renderer::Feature::DepthTest);
     GL::Renderer::enable(GL::Renderer::Feature::FaceCulling);
-    setSize(width, height);
   }
   ~Impl() { LOG(INFO) << "Deconstructing Renderer"; }
 
-  void setSize(int width, int height) {
-    framebufferSize_[0] = width;
-    framebufferSize_[1] = height;
-    colorBuffer_.setStorage(GL::RenderbufferFormat::SRGB8Alpha8,
-                            framebufferSize_);
-    depthBuffer_.setStorage(GL::RenderbufferFormat::R32F, framebufferSize_);
-    objectIdBuffer_.setStorage(GL::RenderbufferFormat::R32UI, framebufferSize_);
-    depthRenderbuffer_.setStorage(GL::RenderbufferFormat::Depth24Stencil8,
-                                  framebufferSize_);
-    framebuffer_ = GL::Framebuffer{{{}, framebufferSize_}};
-    framebuffer_
-        .attachRenderbuffer(GL::Framebuffer::ColorAttachment{0}, colorBuffer_)
-        .attachRenderbuffer(GL::Framebuffer::ColorAttachment{1}, depthBuffer_)
-        .attachRenderbuffer(GL::Framebuffer::ColorAttachment{2},
-                            objectIdBuffer_)
-        .attachRenderbuffer(GL::Framebuffer::BufferAttachment::Depth,
-                            depthRenderbuffer_)
-        .mapForDraw({{0, GL::Framebuffer::ColorAttachment{0}},
-                     {1, GL::Framebuffer::ColorAttachment{1}},
-                     {2, GL::Framebuffer::ColorAttachment{2}}});
-    CORRADE_INTERNAL_ASSERT(
-        framebuffer_.checkStatus(GL::FramebufferTarget::Draw) ==
-        GL::Framebuffer::Status::Complete);
-  }
-
-  inline void renderEnter() {
-    framebuffer_.clear(GL::FramebufferClear::Color |
-                       GL::FramebufferClear::Depth);
-    framebuffer_.clearColor(1, Vector4{});
-    framebuffer_.clearColor(2, Vector4ui{});
-    framebuffer_.bind();
-  }
-
-  inline void renderExit() {}
-
   void draw(RenderCamera& camera, MagnumDrawableGroup& drawables) {
-    renderEnter();
-    camera.getMagnumCamera().setViewport(framebufferSize_);
     camera.draw(drawables);
-    renderExit();
   }
 
   void draw(sensor::Sensor& visualSensor, scene::SceneGraph& sceneGraph) {
     ASSERT(visualSensor.isVisualSensor());
 
     // set the modelview matrix, projection matrix of the render camera;
     sceneGraph.setDefaultRenderCamera(visualSensor);
+    sceneGraph.getDefaultRenderCamera().getMagnumCamera().setViewport(
```

Comment:
Fancy method chaining.

Maybe we can upgrade the "setDefaultRenderCamera" in the SceneGraph.h to
void setDefaultRenderCamera(sensor::Sensor& sensor, Magnum::Vector2i* framebufferSizePtr = nullptr);
In this way, you can hide the getMagnumCamera ()... inside this function with:

if (nullptr != framebufferSizePtr) {
// set the frame buffer size to the default camera;
}

It would be more convenient for the other developers.

Reply:
Another question:
The default camera is a state machine I believe. Once we set its resolution, do we need to restore the default value of it?

Reply:
Made `setDefaultRenderCamera` check to see if the sensor has a rendering target and if it does, set the framebuffer size on the camera

Reply:
> Made `setDefaultRenderCamera` check to see if the sensor has a rendering target and if it does, set the framebuffer size on the camera

I have another question above that you may miss.
Suppose the resolution was set by Sensor A previously. Now Sensor B is going to render something, which does not have a buffer target and would like to render it in the default buffer, would the resolution need to be reset in that case?

Reply:
I guess maybe querying the sensor's spec to figure out the resolution makes more sense?  Also, AFAIK, there currently isn't a default framebuffer.